invenio_app_rdm/records_ui/views/records.py

```diff
@@ -70,10 +70,17 @@ def open(self):
 def record_detail(record=None, files=None, pid_value=None, is_preview=False):
     """Record detail page (aka landing page)."""
     files_dict = None if files is None else files.to_dict()
-
+    record_ui = UIJSONSerializer().serialize_object_to_dict(record.to_dict())
+    if is_preview:
+        # If is preview, only display values if required fields are filled in
+        record_ui['ui']['related_identifiers'] = [
+            rec for rec in record_ui['ui']['related_identifiers']
+            if rec.get('identifier') and rec.get('scheme')
+            and rec.get('relation_type')
+        ]
```

Comment:
Hmm, so although this bug was logged for `related_identifiers` it actually affects `dates`, `funding`, `references`, `alternate_identifiers`...

A small quick "solution" (which isn't really one, but might be good enough for depositors) is to move the switch-case https://github.com/inveniosoftware/react-invenio-deposit/blob/master/src/lib/components/PreviewButton.js#L44 under `case FORM_SAVE_SUCCEEDED` only. This way clicking on preview with an invalid draft doesn't redirect to a breaking preview page. Obviously it's not ideal since one might visit the preview page via its url directly.  

A more complete alternative would be to edit the details macros to account for the possibility of missing subfields. But that's a chore and still error-prone.

Perhaps, easiest would be to also return (in addition to first quick solution) a 404 if `record_ui["errors"]` is not empty when `is_preview` i.e. the preview doesn't exist because a valid draft doesn't exist.  

Alternatives along these lines would work too.

Reply:
I see, I am not sure we can show the 404, since when visiting the preview page via the url directly we don't get the errors. 

As an alternative solution we could maybe add a cleaning function that cleans all the missing required values before previewing, WDYT?

(feel free to ping me in discord, probably easir to discuss IRL)

Reply:
discussed : we will try for a draft service method that just validates a draft so that we can get the `"errors"` in the `if is_preview` conditional. If there are any, we return a 404 page (seemed most appropriate). On the deposit page, we won't even redirect there if there are errors.